### PR TITLE
Extending cryptenroll-tpm2 to support use of systemd-creds

### LINKED ISSUE #31196: cryptenroll: document $NEWPIN env var
### systemd version the issue has been seen with

255

### Used distribution

Arch

### Linux kernel version used

6.7.3-arch1-1

### CPU architectures issue was seen on

None

### Component

systemd

### Expected behaviour you didn't see

Following the documentation: in case there is a wish to use PCRs + PIN to unlock LUKS, the flag to be used is ```--tpm2-with-pin=yes``` and then a pin will be asked. However, this is not handy when generating UKI automatically. Checking the source code for systemd-cryptenroll I have noticed that there is the possibility to use an environment variable NEWPIN to provide the pin, thus enabling automated enrolls... but there is no mention to this variable in the documentation.

Is this env variable expected to go away soon, and thus is not mentioned on the documentation (so, should not be used), or is the documentation outdated? If so, I will prepare a PR to get it fixed.

### Unexpected behaviour you saw

_No response_

### Steps to reproduce the problem

_No response_

### Additional program output to the terminal or log subsystem illustrating the issue

_No response_

### PR DESCRIPTION
**Pull request by flixman**
In the current code base there seems to be no support for automated PIN setting through systemd-creds. It makes use of the NEWPIN environment variable, insecure and mainly aimed at testing. This PR tries to retrieve the new pin through the NEWPIN systemd-cred if available, besides updating the documentation and providing a test for it.
resolves #31196

### CODE REVIEW

File: man/systemd-cryptenroll.xml [hunk 1]
```diff
@@ -549,6 +549,10 @@
         project='mankier'><refentrytitle>tpm2_dictionarylockout</refentrytitle><manvolnum>1</manvolnum></citerefentry>
         commands, respectively.</para>
 
+        <para>Note as well that the new pin can be provided through the NEWPIN environment variable. In case
```
Comment (bluca) [inline man/systemd-cryptenroll.xml]: Don't document the env var here, it's only for testing

File: src/cryptenroll/cryptenroll-tpm2.c [hunk 2]
```diff
@@ -74,55 +75,61 @@ static int get_pin(char **ret_pin_str, TPM2Flags *ret_flags) {
         assert(ret_pin_str);
         assert(ret_flags);
 
-        r = getenv_steal_erase("NEWPIN", &pin_str);
-        if (r < 0)
-                return log_error_errno(r, "Failed to acquire PIN from environment: %m");
+        // try to get the pin from systemd-creds
```
Comment (bluca) [inline src/cryptenroll/cryptenroll-tpm2.c]: /* */ for comments

File: src/cryptenroll/cryptenroll-tpm2.c [hunk 3]
```diff
@@ -74,55 +75,61 @@ static int get_pin(char **ret_pin_str, TPM2Flags *ret_flags) {
...
+        r = read_credential("NEWPIN", (void**) &pin_str, NULL);
```
Comment (bluca) [inline src/cryptenroll/cryptenroll-tpm2.c]: don't call this NEWPIN, but something like tpm2.pin or so

**Review by bluca — CHANGES_REQUESTED** [review]

File: src/cryptenroll/cryptenroll-tpm2.c (hunk 3 above)
Comment (flixman) [reply src/cryptenroll/cryptenroll-tpm2.c]: @bluca I was thinking today morning: does make sense to be sensitive to the env var NEWPIN? Wouldn't be better to just switch it to the cred?

File: man/systemd-cryptenroll.xml (hunk 1 above)
Comment (flixman) [reply man/systemd-cryptenroll.xml]: done

File: src/cryptenroll/cryptenroll-tpm2.c (hunk 2 above)
Comment (flixman) [reply src/cryptenroll/cryptenroll-tpm2.c]: done

File: src/cryptenroll/cryptenroll-tpm2.c (hunk 3 above)
Comment (bluca) [reply src/cryptenroll/cryptenroll-tpm2.c]: env vars are useful for debugging

Comment (flixman) [reply src/cryptenroll/cryptenroll-tpm2.c]: ok, done then.

**Comment by poettering** [comment]
Hmm, the existing code already looks for the `cryptenroll.tpm2-pin`, no?

i.e, it's the 5th argument to ask_password_auto(): a credential name to check.

**Comment by poettering** [comment]
or in other words, everything is already in place, just needs to be documented?

**Comment by poettering** [comment]
@flixman so i looked into this, and the naming of the creds is a bit all over the place. Hence I cleaned this up here:

https://github.com/systemd/systemd/pull/31400

That also adds the docs for the cred you care about. Hence closing this one.

**Pull request closed without merging**